src/CompositeReferenceFieldManager.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\composite_reference;
+
+use Drupal\Component\Utility\NestedArray;
+use Drupal\Core\Entity\EntityFieldManagerInterface;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Field\BaseFieldDefinition;
+use Drupal\Core\Field\FieldConfigInterface;
+use Drupal\Core\Field\FieldDefinitionInterface;
+
+/**
+ * Manager class for composite reference fields.
+ */
+class CompositeReferenceFieldManager implements CompositeReferenceFieldManagerInterface {
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The entity field manager.
+   *
+   * @var \Drupal\Core\Entity\EntityFieldManagerInterface
+   */
+  protected $entityFieldManager;
+
+  /**
+   * CompositeReferenceFieldManager constructor.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager.
+   * @param \Drupal\Core\Entity\EntityFieldManagerInterface $entity_field_manager
+   *   The entity field manager.
+   */
+  public function __construct(EntityTypeManagerInterface $entity_type_manager, EntityFieldManagerInterface $entity_field_manager) {
+    $this->entityTypeManager = $entity_type_manager;
+    $this->entityFieldManager = $entity_field_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getReferencingEntities(EntityInterface $entity): array {
+    $referencing_entities = [];
+
+    $reference_fields = $this->getReferenceFields($entity->getEntityTypeId());
+
+    // Load all entities that have a reference to the given entity.
+    foreach ($reference_fields as $entity_type => $field_names) {
+      $entity_type_storage = $this->entityTypeManager->getStorage($entity_type);
+      $query = $entity_type_storage->getQuery('OR');
+      foreach ($field_names as $field_name) {
+        $query->condition("$field_name.target_id", $entity->id());
+      }
+
+      $ids = $query->execute();
+      if ($ids) {
+        $referencing_entities = array_merge($referencing_entities, $entity_type_storage->loadMultiple($ids));
+      }
+    }
+
+    return $referencing_entities;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function onDelete(EntityInterface $entity, FieldDefinitionInterface $field_definition): void {
+    if ($this->assertFieldDefinition($field_definition)) {
+      $referenced_entities = $entity->get($field_definition->getName())->referencedEntities();
+      /** @var \Drupal\Core\Entity\EntityInterface $referenced_entity */
+      foreach ($referenced_entities as $referenced_entity) {
+        if ($referenced_entity->id() !== $entity->id() && empty($this->getReferencingEntities($referenced_entity))) {
```

Comment:
After testing the solution with having the referencing order: Test 1 -> Test 2 -> Test 3, by deleting Test 1 all the chain was deleted. Logically it is correct, but not sure if this should happen without any warning on the Delete confirmation page as this could potentially wipe out a significant amount of content. I think we could show the affected entity titles that would be deleted on the delete confirmation page or something like that.

Reply:
I don't expect that high chain of deletions...would be a quite a use case and individual sites would have to adopt a different way of deleting.

Reply:
These are assumptions, but still it would be possible. Think about editors who has no idea about this composite configurations, how would they know an impact of their delete action without any message thrown at them?

Reply:
Because of the very nature of this module. It's meant for COMPOSITE entity relations which means nobody even needs ti be aware of the existence of the referenced entities. They are just a storage construct to group multiple fields and that live only with the parent. The site builder uses this module for that. Not for random use cases of "i wanna trigger deletions"..